reactor-netty-http/src/main/java/reactor/netty/http/server/NonSslRedirectHandler.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (c) 2011-Present VMware, Inc. or its affiliates, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package reactor.netty.http.server;
+
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.handler.codec.http.DefaultHttpResponse;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+
+import javax.annotation.Nullable;
+
+import static io.netty.handler.codec.http.HttpHeaderNames.HOST;
+import static io.netty.handler.codec.http.HttpHeaderNames.LOCATION;
+
+/**
+ * @author James Chen
+ */
+final class NonSslRedirectHandler extends ChannelInboundHandlerAdapter {
+
+	private static final String HTTP_PROTOCOL = "http://";
+	private static final String HTTPS_PROTOCOL = "https://";
+	private boolean redirectHttp2Https;
+
+	void enableRedirecting() {
+		this.redirectHttp2Https = true;
+	}
+
+	@Override
+	public void channelRead(ChannelHandlerContext ctx, Object msg) {
+		if (msg instanceof HttpRequest) {
+			ctx.pipeline().remove(this);
+			if (redirectHttp2Https) {
+				HttpRequest request = (HttpRequest) msg;
+				String url = getRequestedUrlInHttps(request);
+				if (url == null) {
+					ctx.close();
+				}
+				else {
+					HttpResponse response = new DefaultHttpResponse(request.protocolVersion(), HttpResponseStatus.PERMANENT_REDIRECT);
```

Comment:
What about returning here a full response? `DefaultFullHttpResponse`?
After writing the response, should we continue with the pipeline?

Reply:
1. `DefaultFullHttpResponse` extends `DefaultHttpResponse` with an additional a http body, and we don't need the body currently.
2. I am sure that we need to continue with the pipeline so that the handler of HttpServer can accept the request and handle it. Otherwise, the handler isn't aware of the request.
https://github.com/reactor/reactor-netty/blob/31664c33b2e78d26b54b6be8e7979b5bec25762d/reactor-netty-http/src/test/java/reactor/netty/http/client/HttpRedirectTest.java#L528-L534

Reply:
The idea of `DefaultFullHttpResponse` is that it combines `DefaultHttpResponse` and `LastHttpContent`, otherwise you have to send two messages `DefaultHttpResponse` and `LastHttpContent` in order to consider the response complete. By default the content is `Unpooled.buffer(0)` so we are OK.

Reply:
What is the use case where the handler should know about the redirect message?

Reply:
1. OK, I will use replace `DefaultHttpResponse` with `DefaultFullHttpResponse`.

2. Continue the pipeline for:
* Other handlers can handle it (especially for metrics-related handlers)
* It's reasonable to pass the HTTP request just because that's what `reactor.netty.http.server.HttpServer#handle` means, and developers can choose whether to just ignore it flexibly.

Reply:
> 2. Continue the pipeline for:
> 
> * Other handlers can handle it (especially for metrics-related handlers)

If we add the handler lazy it will be added before `ReactiveBridge` so no handlers between these two handlers.

> * It's reasonable to pass the HTTP request just because that's what `reactor.netty.http.server.HttpServer#handle` means, and developers can choose whether to just ignore it flexibly.

I cannot understand that.
In the redirect handler, you send a complete response, if you continue with the pipeline then
- the I/O handler will need to distinguish between these two types of requests instead of focusing on serving the necessary data as there was never a redirect.
- the I/O handler cannot do anything to undo the already sent response

Reply:
> we add the handler lazy it will be added before ReactiveBridge so no handlers between these two handlers.

No. Even if we add the handler lazily, there are still some useful handlers between these two handlers (e.g. `reactor.netty.http.server.HttpTrafficHandler` will run before `NonSslRedirectHandler`)

> the I/O handler will need to distinguish between these two types of requests instead of focusing on serving the necessary data as there was never a redirect.

No. the I/O handler doesn't need to distinguish any type of requests and can just handle it like what they usually do. And if they try to send a response, they will just get the exception `AbortedException` and this's reasonable because the handler has redirected HTTP to HTTPS.
https://github.com/reactor/reactor-netty/blob/31664c33b2e78d26b54b6be8e7979b5bec25762d/reactor-netty-http/src/test/java/reactor/netty/http/client/HttpRedirectTest.java#L528-L534

> the I/O handler cannot do anything to undo the already sent response

It's fine for the developer's custom handler cannot undo the already sent response because that's what `redirectHttp2Https` should do (redirect HTTP to HTTPS if SSL enabled). And I don't want to make it complicated (add support for undoing).
If developers really need complicated logic, they should consider writing netty handlers for their cases.

Reply:
> No. Even if we add the handler lazily, there are still some useful handlers between these two handlers (e.g. reactor.netty.http.server.HttpTrafficHandler will run before NonSslRedirectHandler)

We have this pipeline

`HttpTrafficHandler` ...  `NonSslRedirectHandler` (this is only Inbound handler)  `ChannelOperationsHandler` (this is only Inbound handler)

When we read the incoming data the pipeline is the following

`HttpTrafficHandler` -> ... -> `NonSslRedirectHandler` (this is only Inbound handler) -> `ChannelOperationsHandler` (this is only Inbound handler)

so `HttpTrafficHandler` processes the message first and then `NonSslRedirectHandler` is invoked and the last one is `ChannelOperationsHandler`. 

When we write the outgoing data the pipeline is the following
`HttpTrafficHandler` <- ... some outbound handlers (`NonSslRedirectHandler` and `ChannelOperationsHandler` are not invoked when writing)

So there is no handler between `NonSslRedirectHandler` and `ChannelOperationsHandler` that will miss the incoming message.

Reply:
> No. the I/O handler doesn't need to distinguish any type of requests and can just handle it like what they usually do. And if they try to send a response, they will just get the exception `AbortedException` and this's reasonable because the handler has redirected HTTP to HTTPS.
> 
> https://github.com/reactor/reactor-netty/blob/31664c33b2e78d26b54b6be8e7979b5bec25762d/reactor-netty-http/src/test/java/reactor/netty/http/client/HttpRedirectTest.java#L528-L534
> 
> > the I/O handler cannot do anything to undo the already sent response
> 
> It's fine for the developer's custom handler cannot undo the already sent response because that's what `redirectHttp2Https` should do (redirect HTTP to HTTPS if SSL enabled). And I don't want to make it complicated (add support for undoing).
> If developers really need complicated logic, they should consider writing netty handlers for their cases.

I suggest as a first implementation to skip notifying the I/O handler for the redirect response and later if there is a use case to add that functionality and also think about undo.
Wdyt?

Reply:
OK, let's make things easier currently and just skip notifying the I/O handler for the redirect response.
And I will update the PR in 2 hours.